### PR TITLE
feat: add graph subcommand to print dependency graph

### PR DESCRIPTION
**Pull request by AkashKumar7902**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

solves: #263 

#### 2. What is the scope of this PR (e.g. component or file name):

kpm.go
pkg/cmd/cmd_graph.go
pkg/client.go

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [ ] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

**Comment by AkashKumar7902** [comment]
I have implemented my own `findSource` function as the same is not provided by the library.

**Comment by AkashKumar7902** [comment]
@zong-zhe Please take a look

I have implemented two functions here `Union` and `FindSource`. The latter is implemented because the function or similar function is not provided by the libraray. The former is an override of the function provided by the library because I have assumed  graph may be cyclic and error like "vertex already exist" & "edge already exist" needs to be ignored in cyclic graph construction.


Currently the graph is being built bottom-up in recursion , that's why we are getting the graph as a return value. 
We can built the graph in top-down manner but that would require introducing an extra argument to the `downloadDeps` function. The top-down can decrease lines of code.  

instead of using `downloadDeps` for providing two functionality, we can built a clone of `downloadDeps` function with the name say `constructDepGrpah` whose main task would be to construct the graph. 

Please let me know of any changes.

File: pkg/client/client.go [hunk 1]
```diff
@@ -1068,6 +1069,55 @@ func (c *KpmClient) ParseOciOptionFromString(oci string, tag string) (*opt.OciOp
...
+		return err
+	}
+
+	sources, err := FindSource(depGraph)
+	if err != nil {
+		return err
+	}
+
+	// add the root vertex(package name) to the dependency graph.
+	root := fmt.Sprintf("%s@%s", kclPkg.GetPkgName(), kclPkg.GetPkgVersion())
```
Comment (Peefy) [inline pkg/client/client.go:1085]: This graph should be printed to the kpm client logger.

File: pkg/client/client.go [hunk 2]
```diff
@@ -1244,3 +1322,80 @@ func check(dep pkg.Dependency, newDepPath string) bool {
 
 	return dep.Sum == sum
 }
+
+func FindSource[K comparable, T any](g graph.Graph[K, T]) ([]K, error) {
```
Comment (Peefy) [inline pkg/client/client.go:1326]: Move graph related functions to graph.go or dep_graph.go

File: pkg/client/client.go [hunk 3]
```diff
@@ -1244,3 +1322,80 @@ func check(dep pkg.Dependency, newDepPath string) bool {
...
+
+// Union combines two given graphs into a new graph. The vertex hashes in both
+// graphs are expected to be unique. The two input graphs will remain unchanged.
+//
+// Both graphs should be either directed or undirected. All traits for the new
+// graph will be derived from g.
+// 
+// If the same vertex/edge happens to be in both g and h, then an error will not be 
+// thrown as happens in original Union function and successful operation takes place.
+func Union[K comparable, T any](g, h graph.Graph[K, T]) (graph.Graph[K, T], error) {
```
Comment (Peefy) [inline pkg/client/client.go:1353]: Move graph related functions to graph.go or dep_graph.go

Comment (AkashKumar7902) [reply pkg/client/client.go]: Done! 
I have also added unit test

File: pkg/client/client.go (hunk 2 above)
Comment (AkashKumar7902) [reply pkg/client/client.go]: Done !

File: pkg/client/client.go (hunk 1 above)
Comment (AkashKumar7902) [reply pkg/client/client.go]: the graph is getting printed on the cli side https://github.com/kcl-lang/kpm/blob/6fcfb5c0b21f92a79c744bab39d78fd08b646691/pkg/cmd/cmd_graph.go#L77

if you want me to transfer the print logic to client side itself, please let me know.

**Review by Peefy — APPROVED** [review]
LGTM

**Pull request merged**